Restrict the index authorization checks in PIT search

jimczi opened the pull request:
Search requests executed against a PIT do not have to set the target indices explicitly. Instead the list of targeted index is extracted from the PIT id directly. When executed against the search rest layer, the indices of the PIT are copied in the search request in order to fullfill the authorization check when security is enabled. However, this crucial copy operation is omitted when the search request is internally executed from a node within the cluster. Consequently, the authorization check ends up evaluating the request against all indices accessible to the user. To address this performance issue, the proposed change ensures this copy occurs explicitly within the security layer. This modification restricts the authorization check to indices that are part of the PIT, mitigating the performance bug. It is essential to note that even in the presence of this bug, the request does not route to all indices; only those present in the PIT are queried. Furthermore, this issue is not a security vulnerability, as the authorization check continues to limit the list of accessible indices to those permitted for the user.

Closes #99684

jimczi committed:
Restrict the index authorization checks in PIT search

Search requests executed against a PIT do not have to set the target indices explicitly. Instead the list of targeted index is extracted from the PIT id directly.
When executed against the search rest layer, the indices of the PIT are copied in the search request in order to fullfill the authorization check when security is enabled.
However, this crucial copy operation is omitted when the search request is internally executed from a node within the cluster.
Consequently, the authorization check ends up evaluating the request against all indices accessible to the user.
To address this performance issue, the proposed change ensures this copy occurs explicitly within the security layer.
This modification restricts the authorization check to indices that are part of the PIT, mitigating the performance bug.
It is essential to note that even in the presence of this bug, the request does not route to all indices; only those present in the PIT are queried.
Furthermore, this issue is not a security vulnerability, as the authorization check continues to limit the list of accessible indices to those permitted for the user.

Closes #99684

elasticsearchmachine commented:
Pinging @elastic/es-security (Team:Security)

elasticsearchmachine commented:
Pinging @elastic/es-search (Team:Search)

javanna commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java:
```diff
@@ -469,6 +472,26 @@ private void authorizeAction(
+                    listener.onFailure(actionDenied(authentication, authzInfo, action, request, cause));
+                    return;
+                }
+                searchRequest.indices(indices);
```
do we still need to replace the indices or could we just perform the authorization and leave the indices empty? I am thinking that it could be good to enforce that a search request with a pit never specifies indices, but this would break that assumption.

jimczi replied:
++, we don't need to replace the indices I agree. I pushed [89853df](https://github.com/elastic/elasticsearch/pull/103152/commits/89853df0447c35e3cc5620f6edd82de1f05c809e) to resolve the indices directly from the PIT. I also removed the copy in the rest action (this is not needed now since the security layer knows where to look) and moved the request validation to the transport implementation

javanna replied:
I was going to say, you may get some test failures from places that do set the indices. I think for instance ML transform does that.

javanna reviewed (commented):
left a question, this seems like a good direction to me

jimczi committed:
ensure that indices are ordered consistently when extracted from a PIT. Removes the indices from transform pit search requests.

jimczi committed:
Restrict the index authorization checks in PIT search

Search requests executed against a PIT do not have to set the target indices explicitly. Instead the list of targeted index is extracted from the PIT id directly.
When executed against the search rest layer, the indices of the PIT are copied in the search request in order to fullfill the authorization check when security is enabled.
However, this crucial copy operation is omitted when the search request is internally executed from a node within the cluster.
Consequently, the authorization check ends up evaluating the request against all indices accessible to the user.
To address this performance issue, the proposed change ensures this copy occurs explicitly within the security layer.
This modification restricts the authorization check to indices that are part of the PIT, mitigating the performance bug.
It is essential to note that even in the presence of this bug, the request does not route to all indices; only those present in the PIT are queried.
Furthermore, this issue is not a security vulnerability, as the authorization check continues to limit the list of accessible indices to those permitted for the user.

Closes #99684

jimczi committed:
Merge remote-tracking branch 'upstream/main' into search_pit_resolved_indices

jimczi committed:
spotlessApply

jimczi committed:
fix tests that set indices or indices options on a PIT search

jimczi committed:
Handle search options in SQL when a PIT is used

jimczi committed:
spotlessApply

jimczi committed:
Merge remote-tracking branch 'upstream/main' into search_pit_resolved_indices

jimczi committed:
Adapt it tests that set the indices explicitely in a pit search request

jimczi committed:
Adapt it tests that set the indices explicitely in a pit search request

jimczi committed:
Ensure that the RandomizingClient doesn't create invalid request

jimczi committed:
spotlessApply

javanna commented on test/framework/src/main/java/org/elasticsearch/test/client/RandomizingClient.java:
```diff
@@ -84,4 +87,18 @@ public Client in() {
+            if (defaultPreference != null) {
+                setPreference(null);
+            }
+            return super.setPointInTime(pointInTimeBuilder);
```
I guess we should also override setPreference to make sure it throws if point in time is already set, otherwise we rely on order in which the methods are called?

jimczi replied:
I am just checking the default preference here since it is set by the `RandomizingClient` ctr. If a test using this client sets the preference explicitly an error will be thrown during the execution indicating to the author of the test that preference shouldn't be used in this context.

javanna replied:
oh right!

jimczi committed:
Fix remaining tests that set indices on PIT request

jimczi committed:
Merge remote-tracking branch 'upstream/main' into search_pit_resolved_indices

jimczi committed:
spotlessApply

jimczi committed:
fix oobe

jimczi committed:
Merge remote-tracking branch 'upstream/main' into search_pit_resolved_indices

jimczi commented:
The PR is ready for another round of review. I did some changes on the transform indexer too, @hendrikmuhs could you take a look?

jimczi commented:
The PR is ready for another round of review @javanna . @przemekwitek can you take a look at the changes I made to the transform indexer?

javanna reviewed (approved):
LGTM, great work

javanna commented:
@jimczi do you think this would be safe to backport?

jimczi commented:
@javanna I think it's relatively safe but the impact on performance should be minimal so I am not sure if it's needed. The fix is really about reporting the targeted indices correctly so that the tasks and other expert APIs don't expose it wrongly.
